### PR TITLE
CLOUDSTACK-9287 - Fix unique mac address per rVPC router

### PR DESCRIPTION
**Pull request by remibergsma**
This is work by @wilderrodrigues, see PR #1413 It contains important fixes and I think it needs to be included so I send the PR again.

### CODE REVIEW

File: server/src/com/cloud/network/element/VpcVirtualRouterElement.java [hunk 1]
```diff
@@ -466,7 +466,7 @@ public boolean deletePrivateGateway(final PrivateGateway gateway) throws Concurr
             }
         }
 
-        return result > 0 ? true : false;
+        return result == routers.size() ? true : false;
     }
```
Comment (alexandrelimassantana) [inline server/src/com/cloud/network/element/VpcVirtualRouterElement.java]: This line is equivallent to 
```Java
return result == routers.size();
```

please take the ternary operator away

**Comment by DaanHoogland** [comment]
@alexandrelimassantana please make a pull request to the branch that the PR is created from, Remi can include you change in this one by pulling it to his branch.

**Comment by kiwiflyer** [comment]
@remibergsma 

Is this exactly the same PR as the one Wilder closed last week, or does it have some changes in it?

Also, there are some comments from @dsclose added after the old PR was closed regarding some potential gotchas. See https://github.com/apache/cloudstack/pull/1413#issuecomment-205716220

I agree that this (along with the additional fixes proposed above) need to get in to the next release.

- Si

**Comment by dsclose** [comment]
@kiwiflyer 

I'm currently working on CLOUDSTACK-9339 to resolve the issues that I mentioned in #1413 - I don't think this PR breaks anything that currently works so CLOUDSTACK-9339 probably shouldn't hold this PR up.

Dean

**Comment by remibergsma** [comment]
@swill force pushed

**Comment by rhtyd** [comment]
@remibergsma please squash changes to a single commit, and push -f

tag:needlove

**Comment by swill** [comment]
### CI RESULTS

```
Tests Run: 85
  Skipped: 0
   Failed: 1
   Errors: 1
 Duration: 10h 26m 38s
```

**Summary of the problem(s):**
```
FAIL: Test destroy(expunge) Virtual Machine
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/git/cs1/cloudstack/test/integration/smoke/test_vm_life_cycle.py", line 646, in test_09_expunge_vm
    self.assertEqual(list_vm_response,None,"Check Expunged virtual machine is in listVirtualMachines response")
AssertionError: Check Expunged virtual machine is in listVirtualMachines response
----------------------------------------------------------------------
Additional details in: /tmp/MarvinLogs/test_vpc_routers_LJEVIW/results.txt
```

```
ERROR: Test to verify access to loadbalancer haproxy admin stats page
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/git/cs1/cloudstack/test/integration/smoke/test_internal_lb.py", line 854, in tearDown
    raise Exception("Cleanup failed with %s" % e)
Exception: Cleanup failed with Job failed: {jobprocstatus : 0, created : u'2016-05-07T01:17:22+0200', jobresult : {errorcode : 530, errortext : u'Failed to delete network'}, cmd : u'org.apache.cloudstack.api.command.user.network.DeleteNetworkCmd', userid : u'ecdfa0ce-13af-11e6-929e-5254001daa61', jobstatus : 2, jobid : u'7ba837e5-36f2-4467-ab1b-69305fe3c897', jobresultcode : 530, jobresulttype : u'object', jobinstancetype : u'Network', accountid : u'ecdf8a0d-13af-11e6-929e-5254001daa61'}
----------------------------------------------------------------------
Additional details in: /tmp/MarvinLogs/test_network_4LL71G/results.txt
```



**Associated Uploads**

**`/tmp/MarvinLogs/DeployDataCenter__May_06_2016_19_30_38_6T9JP6:`**
* [dc_entries.obj](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1483/tmp/MarvinLogs/DeployDataCenter__May_06_2016_19_30_38_6T9JP6/dc_entries.obj)
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1483/tmp/MarvinLogs/DeployDataCenter__May_06_2016_19_30_38_6T9JP6/failed_plus_exceptions.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1483/tmp/MarvinLogs/DeployDataCenter__May_06_2016_19_30_38_6T9JP6/runinfo.txt)

**`/tmp/MarvinLogs/test_network_4LL71G:`**
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1483/tmp/MarvinLogs/test_network_4LL71G/failed_plus_exceptions.txt)
* [results.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1483/tmp/MarvinLogs/test_network_4LL71G/results.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1483/tmp/MarvinLogs/test_network_4LL71G/runinfo.txt)

**`/tmp/MarvinLogs/test_vpc_routers_LJEVIW:`**
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1483/tmp/MarvinLogs/test_vpc_routers_LJEVIW/failed_plus_exceptions.txt)
* [results.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1483/tmp/MarvinLogs/test_vpc_routers_LJEVIW/results.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1483/tmp/MarvinLogs/test_vpc_routers_LJEVIW/runinfo.txt)


Uploads will be available until `2016-07-07 02:00:00 +0200 CEST`

*Comment created by [`upr comment`](https://github.com/cloudops/upr).*

**Comment by swill** [comment]
I need some code review on this one.  Thanks...

File: server/src/com/cloud/network/element/VpcVirtualRouterElement.java (hunk 1 above)
Comment (DaanHoogland) [inline server/src/com/cloud/network/element/VpcVirtualRouterElement.java]: @alexandrelimassantana This code is not very new, please write a little pull request to that extend. (also the issue that you quite justly pointed out is there in the old code as well and bears no weight on the validity of this PR)

**Comment by DaanHoogland** [comment]
LGTM did a code walk through and I know of a production install this code is running in.

**Comment by swill** [comment]
Can I get one more review on this one?  
@remibergsma are you choosing not to squash because you are not the original author of some of the commits?  If that is the case, I understand and am fine with it.

**Comment by dmabry** [comment]
I tested this in our 4.8 HW Lab and it worked as expected.  I found the MASTER VR for a TEST VPC with a private gateway configured.  I started a ping to the private gateway from an instance.  Finally I ran virsh destroy against r-XX-VM and only dropped a few pings before the private gateway was moved to the new VR.

LGTM

**Pull request merged**